nimbus-consensus/src/lib.rs

```diff
@@ -179,74 +301,16 @@ where
 		relay_parent: PHash,
 		validation_data: &PersistedValidationData,
 	) -> Option<ParachainCandidate<B>> {
-		// Design decision: We will check the keystore for any available keys. Then we will iterate
-		// those keys until we find one that is eligible. If none are eligible, we skip this slot.
-		// If multiple are eligible, we only author with the first one.
-
-		// Get allthe available keys
-		let available_keys =
-			SyncCryptoStore::keys(&*self.keystore, NIMBUS_KEY_ID)
-			.expect("keystore should return the keys it has");
-
-		// Print a more helpful message than "not eligible" when there are no keys at all.
-		if available_keys.is_empty() {
-			warn!(target: LOG_TARGET, "🔏 No Nimbus keys available. We will not be able to author.");
-			return None;
-		}
-
-		let at = BlockId::Hash(parent.hash());
-		// Get `NimbusApi` version.
-		let api_version = self.parachain_client.runtime_api()
-			.api_version::<dyn NimbusApi<B, NimbusId>>(&at)
-			.expect("Runtime api access to not error.");
-
-		if api_version.is_none() {
-			tracing::error!(
-				target: LOG_TARGET, "Could not find `NimbusApi` version.",
-			);
-			return None;
+		let maybe_key = if self.skip_prediction {
+			first_available_key(&*self.keystore)
 		}
-		let api_version = api_version.unwrap();
-
-		// Iterate keys until we find an eligible one, or run out of candidates.
-		// If we are skipping prediction, then we author withthe first key we find.
-		// prediction skipping only really amkes sense when there is a single key in the keystore.
-		let maybe_key = available_keys.into_iter().find(|type_public_pair| {
-
-			// If we are not predicting, just return the first one we find.
-			self.skip_prediction ||
-
-			// Have to convert to a typed NimbusId to pass to the runtime API. Maybe this is a clue
-			// That I should be passing Vec<u8> across the wasm boundary?
-			if api_version >= 2 {
-				self.parachain_client.runtime_api().can_author(
-					&at,
-					NimbusId::from_slice(&type_public_pair.1),
-					validation_data.relay_parent_number,
-					parent,
-				)
-				.expect("Author API should not return error")
-			} else {
-				#[allow(deprecated)]
-				self.parachain_client.runtime_api().can_author_before_version_2(
-					&at,
-					NimbusId::from_slice(&type_public_pair.1),
-					validation_data.relay_parent_number,
-				)
-				.expect("Author API version 2 should not return error")
-			}
-		});
+		else {
+			first_eligible_key::<B, ParaClient>(self.parachain_client.clone(), &*self.keystore, parent, validation_data.relay_parent_number)
+		};
 
-		// If there are no eligible keys, print the log, and exit early.
 		let type_public_pair = match maybe_key {
 			Some(p) => p,
-			None => {
-				info!(
-					target: LOG_TARGET,
-					"🔮 Skipping candidate production because we are not eligible"
-				);
-				return None;
-			}
+			None => { return None; }
 		};
```

Comment:
Why was this log statement removed?

This assignment could also be written more succinctly as 
```rust
let type_public_pair = if let Some(p) = maybe_key {
 p
} else {
 return None
}
```

Reply:
I kind of like the match here. Although, yeah, clippy agrees with you. Anyway, this whole question will be obviated after https://github.com/PureStake/nimbus/issues/16